### PR TITLE
[JITLink][AArch32] Implement Armv5 ldr-pc stubs and use them for all pre-v7 targets

### PR DESCRIPTION
**Pull request by weliveindetail**
This stub type loads an absolute address directly into the PC register. It's the simplest and most compatible way to implement a branch indirection across the entire address space (and probably the slowest as well). It's the ideal fallback for all targets for which we did not (yet) implement a more performant solution.

### CODE REVIEW

**Comment by weliveindetail** [comment]
In RuntimeDyld this is the only stub type available: https://github.com/llvm/llvm-project/blob/release/17.x/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyld.cpp#L988

File: llvm/lib/ExecutionEngine/JITLink/aarch32.cpp [hunk 1]
```diff
@@ -725,6 +725,60 @@ bool GOTBuilder::visitEdge(LinkGraph &G, Block *B, Edge &E) {
...
+      dbgs() << "  Fixing " << G.getEdgeKindName(E.getKind()) << " edge at "
+              << B->getFixupAddress(E) << " (" << B->getAddress() << " + "
+              << formatv("{0:x}", E.getOffset()) << ")\n";
+    });
+    E.setTarget(this->getEntryForTarget(G, E.getTarget()));
+    return true;
+  }
+  case Thumb_Call:
+  case Thumb_Jump24:
+    // BL is never out-of-range and can always be rewritten to BLX inline.
```
Comment (smithp35) [inline llvm/lib/ExecutionEngine/JITLink/aarch32.cpp:764]: As someone not familiar with JITLink this is a bit surprising. A Thumb_Call has a (much pre-v7) shorter range (4 MiB vs 32 Mib) than an Arm_Call. Are Thumb_Call's not permitted to target external symbols?

In theory you would need an interworking stub for a Thumb B to an Arm state symbol.

**Review by smithp35 — COMMENTED** [review]
Not sure I can help too much as I don't think I'm familiar enough with JITLink to know what cases are possible or not. I've left what I can.

File: llvm/lib/ExecutionEngine/JITLink/aarch32.cpp [hunk 2]
```diff
@@ -725,6 +725,60 @@ bool GOTBuilder::visitEdge(LinkGraph &G, Block *B, Edge &E) {
...
+    0x00, 0x00, 0x00, 0x00, // L1: .word S
+};
+
+    // TODO: There is only ARM far stub now. We should add the Thumb stub,
+    // and stubs for branches Thumb - ARM and ARM - Thumb.
+//    writeBytesUnaligned(0xe51ff004, Addr, 4); // ldr pc, [pc, #-4]
+
+Symbol &StubsManager_prev7::createEntry(LinkGraph &G, Symbol &Target) {
+  Block &B = allocStub(G, getStubsSection(G), Armv5LongLdrPc);
+  //LLVM_DEBUG({
```
[the remark this answers is not in the document]
Comment (weliveindetail) [reply llvm/lib/ExecutionEngine/JITLink/aarch32.cpp]: No. The `fixup!` commit cleans this up.

File: llvm/lib/ExecutionEngine/JITLink/aarch32.cpp (hunk 1 above)
Comment (weliveindetail) [reply llvm/lib/ExecutionEngine/JITLink/aarch32.cpp:764]: I concluded it from the comment here: https://github.com/llvm/llvm-project/blob/6009708b4367171ccdbf4b5905cb6a803753fe18/lld/ELF/Thunks.cpp#L219-L222

> In theory you would need an interworking stub for a Thumb B to an Arm state symbol.

Like this one?

https://github.com/llvm/llvm-project/blob/release/17.x/lld/ELF/Thunks.cpp#L859-L866

File: llvm/include/llvm/ExecutionEngine/JITLink/aarch32.h [hunk 3]
```diff
@@ -341,6 +336,37 @@ class GOTBuilder : public TableManager<GOTBuilder> {
   Section *GOTSection = nullptr;
 };
 
+/// Stubs builder emits non-position-independent Arm stubs for pre-v7 CPUs.
+/// These architectures have no MovT/MovW instructions and don't support Thumb2.
+/// BL is the only Thumb instruction that can generate stubs and they can always
+/// be transformed into BLX.
```
[the remark this answers is not in the document]
Comment (weliveindetail) [reply llvm/include/llvm/ExecutionEngine/JITLink/aarch32.h:342]: Yes, sounds reasonable. I don't have a proper hardware to test this either. Thanks!

File: llvm/lib/ExecutionEngine/JITLink/aarch32.cpp [hunk 4]
```diff
@@ -725,6 +725,60 @@ bool GOTBuilder::visitEdge(LinkGraph &G, Block *B, Edge &E) {
...
+  ArrayRef<char> Template(reinterpret_cast<const char *>(Code), Size);
+  return G.createContentBlock(S, Template, orc::ExecutorAddr(), Alignment, 0);
+}
+
+const uint8_t Armv5LongLdrPc[] = {
+    0x04, 0xf0, 0x1f, 0xe5, // ldr pc, [pc,#-4] ; L1
+    0x00, 0x00, 0x00, 0x00, // L1: .word S
+};
+
+    // TODO: There is only ARM far stub now. We should add the Thumb stub,
```
[the remark this answers is not in the document]
Comment (weliveindetail) [reply llvm/lib/ExecutionEngine/JITLink/aarch32.cpp]: This comment was a leftover. Yes, let me implement it!

[the remark this answers is not in the document]
Comment (weliveindetail) [reply llvm/lib/ExecutionEngine/JITLink/aarch32.cpp]: Done with https://github.com/llvm/llvm-project/pull/79082/commits/559606e8b3354482e2e1b759a8e212e8899ac3fa

File: clang/tools/clang-repl/CMakeLists.txt [hunk 5]
```diff
@@ -23,7 +23,7 @@ if(CLANG_PLUGIN_SUPPORT)
   export_executable_symbols_for_plugins(clang-repl)
 endif()
 
-string(TOUPPER ${CMAKE_SYSTEM_PROCESSOR} system_processor)
+string(TOUPPER "${CMAKE_SYSTEM_PROCESSOR}" system_processor)
```
Comment (weliveindetail) [inline clang/tools/clang-repl/CMakeLists.txt:26]: (Unrelated change that is on mainline already)

File: llvm/lib/ExecutionEngine/JITLink/aarch32.cpp [hunk 6]
```diff
@@ -725,6 +725,13 @@ bool GOTBuilder::visitEdge(LinkGraph &G, Block *B, Edge &E) {
   return true;
 }
 
+const uint8_t ArmThumbv5LdrPc[] = {
+    0x78, 0x47,             // bx pc
+    0xfd, 0xe7,             // b #-6 ; Arm recommended sequence to follow bx pc
+    0x04, 0xf0, 0x1f, 0xe5, // ldr pc, [pc,#-4] ; L1
+    0x00, 0x00, 0x00, 0x00, // L1: .word S
+};
```
Comment (weliveindetail) [inline llvm/lib/ExecutionEngine/JITLink/aarch32.cpp:733]: Extended the `v5LdrPc` stub with the Thumb prologue. Thumb B instructions will branch to offset 0 of this block. Thumb BL are be rewritten to BLX and branch to offset 4. Same for Arm branches.

**Review by weliveindetail — COMMENTED** [review]
Thanks for your notes @smithp35. This worked out nicely! A test for Thumb B to Arm interworking is todo, because we need support for `R_ARM_THM_JUMP11` first. I will work on it now. I think it's quite rare though and anyway, this is more than RuntimeDyld ever supported. So, I think we could land it already.

**Comment by smithp35** [comment]
> Thanks for your notes @smithp35. This worked out nicely! A test for Thumb B to Arm interworking is todo, because we need support for `R_ARM_THM_JUMP11` first. I will work on it now. I think it's quite rare though and anyway, this is more than RuntimeDyld ever supported. So, I think we could land it already.

Thanks I'll try and take a look later today, but as I mentioned I'm probably not familiar enough with JITLink to add much value to the code-changes so feel free to land this as an improvement on what is already there.

Just in case you weren't aware. The ABI does not require a stub for R_ARM_THM_JUMP11 for either range extension or interworking, the range is so small it would be quite a challenge to place a stub close enough to it. In practice this makes the 2-byte Thumb B instruction restricted to targets within the same section (or even same function), which a code-generator won't change-state. An assembly programmer might, but if they do then they have to do the interworking manually.

It is possible in Thumb-1 to have a 4-byte B instruction that uses R_ARM_THM_JUMP24 that will need a stub from Thumb to Arm.

This is in Call and Jump Relocations https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst
```
A linker may use a veneer (a sequence of instructions) to implement the
relocated branch if the relocation is one of R_ARM_PC24, R_ARM_CALL,
R_ARM_JUMP24, (or, in Thumb state, R_ARM_THM_CALL, R_ARM_THM_JUMP24, or
R_ARM_THM_JUMP19) and:
```

**Review by lhames — COMMENTED** [review]
LGTM. I love the detail in the testing here -- thanks @weliveindetail!
